quinn-proto/src/connection/mod.rs

```diff
@@ -500,9 +541,47 @@ where
                 coalesce = false;
                 None
             } else {
-                Some(self.populate_packet(space_id, &mut builder.buffer))
+                Some(self.populate_packet(
+                    space_id,
+                    &mut builder.buffer,
+                    buf_capacity,
+                    &mut concat_datagram,
+                ))
             };
 
+            // Don't use GSO if we reached the maximum amount of datagrams.
+            // Also don't allow it for other spaces than Data until the impact
+            // is better understood.
+            if num_datagrams == MAX_DATAGRAMS - 1 || space_id != SpaceId::Data {
+                concat_datagram = false;
+            }
+
+            if concat_datagram {
+                // Pad the packet to make it suitable for sending with GSO
+                // which will always send the maxium PDU.
+                // TODO: Can this `tag_len` be taken from the builder?
+                // This is currently copied from the `finish_packet` path.
```

Comment:
Stashing `tag_len` in the builder for easy access is fine. Or maybe it would be simpler of we redefined `min_size` to include the tag?

Reply:
> Or maybe it would be simpler of we redefined min_size to include the tag?

I think this would just require the tag to be known in another place then.

Storing it in the builder sounds good to me. I just wasn't sure about all the places where it is created, and whether it's safe to store it there.

Reply:
> I think this would just require the tag to be known in another place then.

It might reduce it to places where we're already fetching it anyway.